File: spring-web-java/src/main/java/spring/web/java/global/exception/ControllerAdvice.java
package spring.web.java.global.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.validation.FieldError;
import org.springframework.web.HttpRequestMethodNotSupportedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import lombok.RequiredArgsConstructor;

@RestControllerAdvice
@RequiredArgsConstructor
public class ControllerAdvice {

	@ExceptionHandler(DomainException.class)
	public ProblemDetail domain(DomainException exception) {
		ProblemDetail problemDetail = ProblemDetail.forStatus(exception.getHttpStatus());
		problemDetail.setTitle(exception.getResponseMessage().getTitle());
		problemDetail.setDetail(exception.getResponseMessage().getDetail());

		return problemDetail;
	}

	@ExceptionHandler(HttpRequestMethodNotSupportedException.class)
	public ProblemDetail methodNotAllowed(HttpRequestMethodNotSupportedException exception) {
		ProblemDetail problemDetail = ProblemDetail.forStatus(HttpStatus.METHOD_NOT_ALLOWED);
		problemDetail.setTitle(HttpStatus.METHOD_NOT_ALLOWED.getReasonPhrase());
		problemDetail.setDetail(exception.getMessage());

		return problemDetail;
	}

	@ExceptionHandler(MethodArgumentNotValidException.class)
	public ProblemDetail badRequest(MethodArgumentNotValidException exception) {
		ProblemDetail problemDetail = ProblemDetail.forStatus(HttpStatus.BAD_REQUEST);
		problemDetail.setTitle(HttpStatus.BAD_REQUEST.getReasonPhrase());
		problemDetail.setDetail(
			"%s: %s".formatted(
				exception.getMessage(),
				exception.getBindingResult().getFieldErrors().stream()
					.map(FieldError::getField)
					.toList()
			)
		);

		return problemDetail;
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/global/exception/ControllerAdvice.kt
package spring.web.kotlin.global.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.ErrorResponse
import org.springframework.web.HttpRequestMethodNotSupportedException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.resource.NoResourceFoundException

@RestControllerAdvice
class ControllerAdvice {

    @ExceptionHandler(DomainException::class)
    fun handleDomainException(exception: DomainException) =
        ProblemDetail.forStatus(exception.httpStatus).apply {
            title = exception.responseMessage.title
            detail = exception.responseMessage.detail
        }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleInvalidRequestBody(exception: HttpMessageNotReadableException) =
        ProblemDetail.forStatus(HttpStatus.BAD_REQUEST).apply {
            title = HttpStatus.BAD_REQUEST.reasonPhrase
            detail = exception.message
        }

    @ExceptionHandler(NoResourceFoundException::class, HttpRequestMethodNotSupportedException::class)
    fun handleResourceNotFound(errorResponse: ErrorResponse) = errorResponse.body

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(exception: MethodArgumentNotValidException) =
        exception.body.apply {
            setProperty("fieldErrors", exception.bindingResult.fieldErrors.associate { it.field to it.defaultMessage })
        }
}
